chain/network/src/routing/graph_v2/mod.rs

```diff
@@ -0,0 +1,668 @@
+            }
+        }
+        // After this point, we know that the DistanceVector message is valid
+        if !consistent {
+            return None;
+        }
+
+        // Now, prune any advertised routes which go through the local node; it doesn't make
+        // sense to forward a message to a neighbor who will send it back to us
+        let local_node_id = self.edge_cache.get_local_node_id() as usize;
+        for id in 0..self.edge_cache.max_id() {
+            if first_step[id] == local_node_id as i32 && id != local_node_id {
+                advertised_distances[id] = -1;
+            }
+        }
+
+        Some(advertised_distances)
+    }
+
+    /// Accepts a validated DistanceVector and its `advertised_distances`.
+    /// Updates the status of the direct connection between the local node and the direct peer.
+    /// If the peer can be used for forwarding, stores the advertised distances.
+    /// Returns true iff the distances are stored.
+    fn store_validated_peer_distances(
+        &mut self,
+        distance_vector: &network_protocol::DistanceVector,
+        mut advertised_distances: Vec<i32>,
+    ) -> bool {
+        let local_node_id = self.edge_cache.get_local_node_id() as usize;
+
+        // A direct peer's distance vector which advertises an indirect path to the local node
+        // is outdated and can be ignored.
+        if advertised_distances[local_node_id] > 1 {
+            // TODO(saketh): We could try to be more clever here and do some surgery on the tree
+            // to replace the indirect path and speed up convergence of the routing protocol.
+            return false;
+        }
+
+        // Look in the spanning tree for the direct edge between the local node and the root
+        let tree_edge = distance_vector.edges.iter().find(|edge| {
+            edge.contains_peer(&self.config.node_id) && edge.contains_peer(&distance_vector.root)
+        });
+
+        // If the tree has more recent state for the direct edge, replace the local state
+        if let Some(tree_edge) = tree_edge {
+            self.local_edges
+                .entry(distance_vector.root.clone())
+                .and_modify(|local_edge| {
+                    if tree_edge.nonce() > local_edge.nonce() {
+                        *local_edge = tree_edge.clone();
+                    }
+                })
+                .or_insert(tree_edge.clone());
+        }
+
+        // Without a direct edge, we cannot use the distances advertised by the peer
+        let Some(local_edge) = self.local_edges.get(&distance_vector.root) else {
+            return false;
+        };
+        if local_edge.edge_type() == EdgeState::Removed {
+            return false;
+        }
+
+        // If the spanning tree doesn't already include the direct edge, add it
+        let mut spanning_tree = distance_vector.edges.clone();
+        if tree_edge.is_none() {
+            debug_assert!(advertised_distances[local_node_id] == -1);
```

Comment:
I think if this condition is true you want to also return false. The debug assert will not panic in an production build.